collector/lib/ConfigLoader.cpp

```diff
@@ -8,114 +14,520 @@ namespace collector {
+    }
+    case FieldDescriptor::TYPE_ENUM: {
+      auto enum_name = node.as<std::string>();
+
+      // We assume enum definitions use UPPER_CASE nomenclature, so
+      std::transform(enum_name.begin(), enum_name.end(), enum_name.begin(), [](char c) {
+        return std::toupper(c);
+      });
+
+      const EnumDescriptor* descriptor = field->enum_type();
+      const EnumValueDescriptor* value = descriptor->FindValueByName(enum_name);
+      if (value == nullptr) {
+        ParserError err;
+        err << file_ << ": Invalid enum value '" << enum_name << "' for field " << name;
+        return {{err}};
+      }
+      msg->GetReflection()->SetEnumValue(msg, field, value->number());
+    } break;
+
+    case FieldDescriptor::TYPE_MESSAGE:
+    case FieldDescriptor::TYPE_GROUP: {
+      ParserError err;
+      err << "Unexpected type: " << field->type_name();
+      return {{err}};
+    }
+  }
+
+  return {};
+}
+
+ParserResult ParserYaml::ParseArray(google::protobuf::Message* msg, const YAML::Node& node,
+                                    const google::protobuf::FieldDescriptor* field) {
+  using namespace google::protobuf;
+
+  // mapping for repeated fields:
+  // https://protobuf.dev/reference/cpp/api-docs/google.protobuf.message/#Reflection.GetRepeatedFieldRef.details
+  switch (field->cpp_type()) {
+    case FieldDescriptor::CPPTYPE_INT32:
+      return ParseArrayInner<int32>(msg, node, field);
+
+    case FieldDescriptor::CPPTYPE_UINT32:
+      return ParseArrayInner<uint32_t>(msg, node, field);
+
+    case FieldDescriptor::CPPTYPE_INT64:
+      return ParseArrayInner<int64_t>(msg, node, field);
+
+    case FieldDescriptor::CPPTYPE_UINT64:
+      return ParseArrayInner<uint64_t>(msg, node, field);
+
+    case FieldDescriptor::CPPTYPE_DOUBLE:
+      return ParseArrayInner<double>(msg, node, field);
+
+    case FieldDescriptor::CPPTYPE_FLOAT:
+      return ParseArrayInner<float>(msg, node, field);
+
+    case FieldDescriptor::CPPTYPE_BOOL:
+      return ParseArrayInner<bool>(msg, node, field);
+
+    case FieldDescriptor::CPPTYPE_ENUM:
+      return ParseArrayEnum(msg, node, field);
+
+    case FieldDescriptor::CPPTYPE_STRING:
+      return ParseArrayInner<std::string>(msg, node, field);
+
+    case FieldDescriptor::CPPTYPE_MESSAGE: {
+      return {{"Unsupport repeated type MESSAGE"}};
+    } break;
+
+    default: {
+      ParserError err;
+      err << "Unknown type " << field->type_name();
+      return {{err}};
+    }
+  }
+}
+
+ParserError ParserYaml::WrapError(const std::exception& e) {
+  ParserError err;
+  err << file_ << ": " << e.what();
+  return err;
+}
+
+template <typename T>
+std::variant<T, ParserError> ParserYaml::TryConvert(const YAML::Node& node) {
+  try {
+    return node.as<T>();
+  } catch (YAML::InvalidNode& e) {
+    return WrapError(e);
+  } catch (YAML::BadConversion& e) {
+    return WrapError(e);
   }
+}
+
+std::string ParserYaml::SnakeCaseToCamel(const std::string& s) {
+  std::string out;
```

Comment:
```suggestion
  std::string out;
  out.reserve(s.size());
```
Might avoid reallocation in some cases.